CDP4Orm/AutoGenDao/ActualFiniteStateDao.cs

```diff
@@ -226,6 +222,62 @@ public virtual bool Write(NpgsqlTransaction transaction, string partition, CDP4C
             return this.AfterWrite(beforeWrite, transaction, partition, actualFiniteState, container);
         }
 
+        /// <summary>
+        /// Insert a new database record, or updates one if it already exists from the supplied data transfer object.
+        /// This is typically used during the import of existing data to the Database.
+        /// </summary>
+        /// <param name="transaction">
+        /// The current <see cref="NpgsqlTransaction"/> to the database.
+        /// </param>
+        /// <param name="partition">
+        /// The database partition (schema) where the requested resource will be stored.
+        /// </param>
+        /// <param name="actualFiniteState">
+        /// The actualFiniteState DTO that is to be persisted.
+        /// </param>
+        /// <param name="container">
+        /// The container of the DTO to be persisted.
+        /// </param>
+        /// <returns>
+        /// True if the concept was successfully persisted.
+        /// </returns>
+        public virtual bool Upsert(NpgsqlTransaction transaction, string partition, CDP4Common.DTO.ActualFiniteState actualFiniteState, CDP4Common.DTO.Thing container = null)
```

Comment:
Could this method ever have side effects like `BeforeUpsert` and `AfterUpsert` (i.e. are they defined in the base class/interface?

Reply:
No, that is indeed not the intention.

Reply:
Sorry, I was probably editing the question while you added the answer. What I meant to ask is if there's any way to enforce that `BeforeUpsert` and `AfterUpsert` methods are not added to the base class/interface at a later time, since this would probably ruin the data migration procedure flow.

Reply:
We added ```/// This is typically used during the import of existing data to the Database.``` to the comment of each Upsert and RawDelete method. I think that should be enough for now.